ktlint-rule-engine-core/src/main/kotlin/com/pinterest/ktlint/rule/engine/core/api/ASTNodeExtension.kt

```diff
@@ -555,3 +573,9 @@ public fun ASTNode.replaceWith(node: ASTNode) {
 public fun ASTNode.remove() {
     treeParent.removeChild(this)
 }
+
+public fun ASTNode.getPrevSiblingIgnoringWhitespaceAndComments(): ASTNode? =
```

Comment:
In `SpacingBetweenDeclarationsWithCommentsRule` the PSI version of `getPrevSiblingIgnoringWhitespaceAndComments` is used. I was trying to see whether I could refactor the code below, to not use `psi`:
```
            val declaration = node.parent as? KtDeclaration ?: return
            val isTailComment = node.startOffset > declaration.startOffset
            if (isTailComment || declaration.getPrevSiblingIgnoringWhitespaceAndComments() !is KtDeclaration) return
```
I could not (easily) find an angle how to replace `KtDeclaration` with an `ASTNode` implementation. Can you give me a pointer how to approach that? You don't need to implement it, but I would learn to know for myself how I could do that.

Reply:
The question was indeed, how should I have gotten the idea to use `KtTokenSets.DECLARATION_TYPES` instead of `is KtDeclaration`? Could you derive that somehow from the Psi library, or did you find this accidentally while exploring the lib?